[MLIR][WASM] Control flow, conversion and comparison in Wasm importer

lforg37 opened the pull request:
This is the following of PR #154452.

It extend Wasm binary to Wasm SSA importer with support of control flow operations, comparison operations and conversion operations.

lforg37 committed:
[mlir][wasm] Add support for comparison ops in importer

---------

Co-authored-by: Ferdinand Lemaire <ferdinand.lemaire@woven-planet.global>
Co-authored-by: Jessica Paquette <jessica.paquette@woven-planet.global>

lforg37 committed:
[mlir][wasm] Support for conversion ops in Wasm importer

---------

Co-authored-by: Ferdinand Lemaire <ferdinand.lemaire@woven-planet.global>
Co-authored-by: Jessica Paquette <jessica.paquette@woven-planet.global>

lforg37 committed:
[mlir][wasm] Support for control flow related ops in Wasm importer

Add label level tracking system to the importer.
Add import for all the control flow related ops.

---------

Co-authored-by: Ferdinand Lemaire <ferdinand.lemaire@woven-planet.global>
Co-authored-by: Jessica Paquette <jessica.paquette@woven-planet.global>

lforg37 commented:
@joker-eph could you have a look?

lforg37 commented on mlir/lib/Target/Wasm/TranslateFromWasm.cpp:
```diff
@@ -701,17 +853,41 @@ inline parsed_inst_t ExpressionParser::parseSpecificInstruction(OpBuilder &) {
+  auto nextFrameIdx = getNextFrameIndex();
   size_t stackSize = size();
-  for (size_t idx = 0; idx < stackSize; idx++) {
+  for (size_t idx = 0; idx < stackSize;) {
```
Yes it seems so. I'll investigate, as it seems something got lost here. 
Basically, I recalled the lambdas were responsible to update the index but this is not the case here.
Thanks for noticing it.

Commit:
[WASM][MLIR] Fix dump method in Wasm importer

flemairen6 commented:
@jpienaar since @joker-eph seems to be off, can you have a look at this PR? Thanks a lot :)

joker-eph commented:
The change is really too wasm-specific for me, it would be great if someone else who is interested in these could review.
I saw that @MattPD was reviewing, I thought they would be able to approve this?

MattPD commented:
@lforg37 Thanks, it's a matter of finding some spare cycles (limited bandwidth at the moment)

jpienaar reviewed (commented):
Just some comments from first scan. If I recall from the ODM, the yaml with binary strings is the best currently given import support, correct?

Commit:
Apply review comments

lforg37 commented on mlir/lib/Target/Wasm/TranslateFromWasm.cpp:
```diff
@@ -280,11 +308,102 @@ class ExpressionParser {
     }
   }
 
+  struct NestingContext {
```
Done. I also renamed the struct in an attempt to make its role more explicit.

kateinoigakukun commented on mlir/lib/Target/Wasm/TranslateFromWasm.cpp:
```suggestion
    if (type.id >= symbols.moduleFuncTypes.size())
```

lforg37 replied:
Thanks for spotting this

kateinoigakukun commented on mlir/lib/Target/Wasm/TranslateFromWasm.cpp:
Would you mind clarifying why `SymbolTable::setSymbolVisibility` helper didn't work for this case, and leave comments if needed?

lforg37 replied:
I think it is because we rely on the fact that the visibility attribute stream should be present somewhere else, and `setSymbolVisibility` remove the attribute for public symbols.

I'll investigate to see where we have this assumption to see if we can instead have follow the usual assumption that no attribute means public visibility.

lforg37 replied:
It seems the only reason we are doing it is to enforce explicit presence of the public keyword in the printer of function op.
The rationale was that public is non default for Wasm function (as it requires an export).

If this is something we want to keep, we need to enforce the existence of the visibility symbol in the verifier of wasmssa function op I think.

Otherwise, we can just have the absent attribute mean public.

Do you have any opinion on this @kateinoigakukun or @jpienaar  ?

kateinoigakukun replied:
I don't have much opinion on MLIR side but it would be nice to add some comments explaining that

lforg37 replied:
OK. I will wait for @jpienaar's opinion then and either add comments or revert to using `SymbolTable::setSymbolVisibility` depending onn the result.

lforg37 replied:
My apologies, the question was unclear. My concern was that if we use the `public` vs `nested` in the textual assembly with `nested` being the default, it might be confusing to user used to e.g. the `func` dialect where `public` is the default.

I will update the doc and add the commit to this PR.

kateinoigakukun reviewed (approved):
I reviewed Wasm-specific changes, and it looks good to me :)

lforg37 commented on mlir/lib/Target/Wasm/TranslateFromWasm.cpp:
```diff
@@ -280,11 +308,102 @@ class ExpressionParser {
+                                               TypeIdxRecord type) {
+    if (type.id > symbols.moduleFuncTypes.size())
+      return emitError(*currentOpLoc,
+                       "Type index references nonexistent type: ")
```
Changed it, and added a test for it.

lforg37 commented on mlir/lib/Target/Wasm/TranslateFromWasm.cpp:
```diff
@@ -701,17 +853,41 @@ inline parsed_inst_t ExpressionParser::parseSpecificInstruction(OpBuilder &) {
+  for (size_t idx = 0; idx < stackSize; ++idx) {
     size_t actualIdx = stackSize - 1 - idx;
+    while (nextFrameIdx && (nextFrameIdx->second > actualIdx)) {
+      llvm::dbgs() << "  --------------- Frame (" << nextFrameIdx->first
```
This is a dump method, should we use LDBG here also? Comment was already resolved, but I am not sure if that is the reason why.

lforg37 committed:
[mlir][Wasm] Wasm like symbol visibility for WasmSSA dialect

Wasm has the concept of export operation that makes a symbol available for use in other modules.
By default an operation has nested visibility only.

This commit changes the way visibility is represented in `wasmssa` dialect to align with Wasm semantics.

Operation that defines "exportable" symbols (memory, global, functions or table) can have an `UnitAttr` that signal they should be exported (and have public visibility). 

When they do not have this attribute, visibility is nested.

---------

Co-authored-by: Luc Forget <luc.forget@woven.toyota>

joker-eph commented:
@flemairen6 feel free to follow [this process](https://llvm.org/docs/DeveloperPolicy.html#obtaining-commit-access) to request commit access to be able to merge your PR without my involvement.
